add drop command and replace the old spawn command

Berthalamew opened the pull request.

Berthalamew committed:
misc changes

- adjust logic for GetVKeyCodeString
- adjust logic for custom_settings_real_check
- remove configureXinput function and all xinput patching logic (not needed anymore)
- renamed current_shared_datum to absolute_index in scenario_tags_load_process_shared_tags
- add observer_try_and_get_camera
- rename cseries_debug_initialize to cseries_windows_debug_initialize
- adjust includes
- adjust _unit_datum
- rename emblem enums
- remove FLOATING_POINT_ENV_ACCESS
- remove isFloat from utils
- remove hash logic

Berthalamew committed:
use an array instead of a vector for custom variants

Berthalamew committed:
add append_print member function to static string classes

- add memmove_guarded to cseries

Berthalamew committed:
add drop command and replace the old spawn command

nukeulater commented on xlive/Blam/Engine/cseries/cseries.h:
```diff
@@ -210,7 +212,7 @@ int32 vsprintf(char* buffer, size_t size, const char* format, va_list va_args);
 * size_t max_count was added in vista to this function call
 */
-int32 vsnprintf(char* buffer, size_t size, size_t max_count, const char* format, va_list va_args);
+int32 vsnprintf(char* buffer, size_t size, size_t max_count, const char* format, char* ap);
```
vsnprintf is a library function? 

https://cplusplus.com/reference/cstdio/vsnprintf/

Berthalamew replied:
Bungie would end up re-using the same CRT function names for their cseries wrappers for halo 2.
This is intentional.

nukeulater replied:
then why did you change the parameter?

Berthalamew replied:
took a look at halo 1 and they seem to pass the va_list as a char*

Berthalamew committed:
move scenario version check outside of map repository

- add extra errors and asserts
- add assert_valid_real_vector3d_axes2 macro
- fix assert_valid_real_normal3d

nukeulater commented on xlive/Blam/Engine/objects/objects.cpp:
```diff
@@ -1216,4 +1329,97 @@ static int16 __cdecl internal_object_get_markers_by_string_id(datum object_index
+			object_header_datum* object_header = (object_header_datum*)datum_try_and_get(object_header_data_get(), g_object_override_data[0].object_index);
+			if (object_header)
+			{
+				if (TEST_BIT(0, object_header->type))
```
this will always fail/be false?

twist84 replied:
Berth this doesn't look like a typical bit test, its only doing `1 << object_header->type`

twist84 replied:
On further inspection it looks like this is just an inlined version of `object_get` or `object_get_and_verify_type`

Berthalamew replied:
changed to use the object_get_and_verify_type function

nukeulater commented on xlive/Blam/Engine/game/cheats.cpp:
```diff
@@ -82,13 +130,203 @@ void __cdecl ice_cream_flavor_stock(const e_skull_type skull)
-	DETOUR_ATTACH(p_ice_cream_flavor_available, Memory::GetAddress<ice_cream_flavor_available_t>(0xBD114, 0xAC2A2), ice_cream_flavor_available);
-	DETOUR_ATTACH(p_ice_cream_activation, Memory::GetAddress<ice_cream_activation_t>(0xBD137, 0xAC2C5), ice_cream_flavor_stock);
+	const datum tag_index = tag_loaded(group.group, name);
+	if (tag_index == -1)
```
== NONE

Berthalamew replied:
changed

Berthalamew committed:
fix logic in object_get_override_index

Berthalamew committed:
cheat re-org

- remove function pointers in ice_cream_flavor_stock and replace with proper function calls
- re-organize some other files and separate private and public functions
- turn off JMC
